### PR TITLE
Python Quality of Life Updates

### PR DESCRIPTION
**Pull request by ddimaria**
Addresses: 
* https://github.com/quadratichq/quadratic/issues/1056
* https://github.com/quadratichq/quadratic/issues/627
* https://github.com/quadratichq/quadratic/issues/53
* https://github.com/quadratichq/quadratic/issues/567

- [x] Inspect return types in the console (add icon next to return line, hide icon with unsaved changes)
- [x] Support rust types -> python types and python types -> rust types
- [x] Output series when appropriate
- [x] Fix cases where the editor hangs if a TS error occurs when calculating python
- [x] Support the Pyright LSP (in JS in a web worker) for auto completion, hover and signature
- [x] Autogenerate python typshed stubs of Quadratic code and supported third-party libraries
- [x] Fix deltaDecorations bug in useEditorCellHighlights.ts
- [x] Enhance regex python parser to handle the various forms of cell/cells/c/getCell/getCells
- [x] Utilize Pyright diagnostics in Monaco
- [x] Handle NoneType returns in the UI 
- [x] Pyodide within a web worker now working in tests
- [x] Fixs stop-python-execution regression

Known Issues
- [x] AST parser sometimes returns the incorrect return line number
- [ ] Top-level async/await supported in Pyodide, but Pyright complains
- [ ] Hover overlay is not respecting markdown (see autocomplete for correct markdown support)
- [ ] Regex parser for Quadratic function highlights is naive and should be replaced with an AST parsed version

### CODE REVIEW

**Comment by ddimaria** [comment]
> * retain the return info, currently once you start typing the return info disappears but appears if you undo, return info should be retained in the console until the next execution

We had some Slack discussions on this.  I believe we're going to stick with the current functionality.

**Comment by jimniels** [comment]
Actions that fetch data over the network cause re-rendering the line number. Like you hit "run", it draws the line number, then re-renders it causing the number to jump.

https://github.com/quadratichq/quadratic/assets/1316441/bc3d0b70-ce66-46a4-b01a-3a46e893c8eb

Feels like we shouldn't render the data until we know exactly what it is (and what line number it's on)

We seem to already be doing this with the in-editor return arrow. It only renders once the operation is complete, whereas the "Line XX returned..." renders first then re-renders once complete

https://github.com/quadratichq/quadratic/assets/1316441/6569569b-5b2b-448e-85d8-d3e7c3cf9b7c

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 1]
```diff
@@ -118,6 +128,14 @@ export const CodeEditor = () => {
...
   }, [updateCodeCell]);
 
+  useEffect(() => {
+    const updateDiagnostics = (e: Event) => setDiagnostics((e as CustomEvent).detail.diagnostics);
+    window.addEventListener('python-diagnostics', updateDiagnostics);
+    return () => {
+      window.removeEventListener('python-diagnostics', updateDiagnostics);
+    };
+  }, [updateCodeCell]);
+
```
Comment (davidkircos) [inline quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx:138]: Is this still used now that codeRunResult stores the types?

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 2]
```diff
@@ -36,6 +41,7 @@ export const CodeEditor = () => {
   const [consoleHeight, setConsoleHeight] = useState<number>(200);
   const [showSaveChangesAlert, setShowSaveChangesAlert] = useState(false);
   const [editorContent, setEditorContent] = useState<string | undefined>(codeString);
+  const [diagnostics, setDiagnostics] = useState<Diagnostic[]>([]);
```
Comment (davidkircos) [inline quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx:44]: Seems like this is only used in the `CodeEditorBody` should we move the state to that component?

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx (hunk 1 above)
Comment (ddimaria) [reply quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx:138]: It's used, but we currently don't do anything with it.  I'll comment it out with notes (as I did in other places regarding diagnostics).  I do think we should come back to diagnostics at some point.

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx [hunk 3]
```diff
@@ -1,40 +1,59 @@
...
 import { FormulaLanguageConfig, FormulaTokenizerConfig } from './FormulaLanguageModel';
-import { provideCompletionItems as provideCompletionItemsPython } from './PythonLanguageModel';
+import {
+  provideCompletionItems as provideCompletionItemsPython,
+  provideHover as provideHoverPython,
+  provideSignatureHelp as provideSignatureHelpPython,
+} from './PythonLanguageModel';
 import { QuadraticEditorTheme } from './quadraticEditorTheme';
 import { useEditorCellHighlights } from './useEditorCellHighlights';
+// import { useEditorDiagnostics } from './useEditorDiagnostics';
```
Comment (davidkircos) [inline quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx:19]: unused import

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx (hunk 2 above)
Comment (ddimaria) [reply quadratic-client/src/ui/menus/CodeEditor/CodeEditor.tsx:44]: Most of the diagnostic code has been commented out with notes attached.  I'll do the same with this.

File: quadratic-client/src/ui/menus/CodeEditor/useLanguageServer.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    endLineNumber: position.lineNumber,
+    endColumn: position.column,
+  });
+  console.log('textUntilPosition', textUntilPosition);
+  var match = textUntilPosition.match(/cel/);
+  if (!match) {
+    return { suggestions: [] };
+  }
+  // var word = model.getWordUntilPosition(position);
+  // var range = {
```
Comment (davidkircos) [inline quadratic-client/src/ui/menus/CodeEditor/useLanguageServer.ts:20]: keep?

File: quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx [hunk 5]
```diff
@@ -44,35 +63,47 @@ export const CodeEditorBody = (props: Props) => {
...
+    monaco.languages.registerHoverProvider('python', { provideHover: provideHoverPython });
+
+    // load the document in the python language server
+    pyrightWorker?.openDocument({
+      textDocument: { text: editorRef.current?.getValue() ?? '', uri, languageId: 'python' },
+    });
+
+    registered = true;
+
+    setDidMount(true);
```
Comment (davidkircos) [inline quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx:105]: how is registered and didMount different?

Comment (ddimaria) [reply quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx:105]: This doesn't work for the monaco singleton.  It's remounts each time.

Comment (ddimaria) [reply quadratic-client/src/ui/menus/CodeEditor/CodeEditorBody.tsx:105]: I kept it as it's being used by another part of the component.

**Review by davidkircos — APPROVED** [review]

**Pull request merged**